WMF Framework/Event Platform Client/EventPlatformClientLibrary.swift

```diff
@@ -0,0 +1,605 @@
+     * For persistent storage
+     */
+    private var storage_manager: StorageManager?
+
+    /**
+     * For handling HTTP requests
+     */
+    private var network_manager: NetworkManager?
+
+    /**
+     * Store events until the library is finished initializing
+     *
+     * The EPC library makes an HTTP request to a remote stream configuration service for information
+     * about how to evaluate incoming event data. Until this initialization is complete, we store any incoming
+     * events in this buffer.
+     */
+    private var input_buffer = [(stream: String, schema: String, data: [String: Any], domain: String?)]()
+
+    /**
+     * Cache of "in sample" / "out of sample" determination for each stream
+     *
+     * The process of determining only has to happen the first time an event is logged to a stream for
+     * which stream configuration is available. All other times `in_sample` simply returns the cached
+     * determination.
+     */
+    private var cache: [String: Bool] = [String: Bool]()
+
+    // MARK: - Methods
+
+    private init() {
+        EVENTGATE_URI = "https://intake-analytics.wikimedia.org/v1/events"
+        CONFIG_URI = "https://meta.wikimedia.org/w/api.php?action=streamconfigs&format=json"
+        ISO8601_FORMATTER = ISO8601DateFormatter()
+        ENABLED = true
+    }
+
+    /**
+     * Tell which storage manager EPC should use
+     *
+     * ERROR: Method cannot be declared public because its parameter uses an internal type
+     */
+    public func set_storage_manager(_ sm: StorageManager) {
+        storage_manager = sm
+        recall_buffer()
+    }
+
+    /**
+     * Tell which network manager EPC should use
+     *
+     * ERROR: Method cannot be declared public because its parameter uses an internal type
+     */
+    public func set_network_manager(_ nm: NetworkManager) {
+        network_manager = nm
+        configure()
+    }
+
+    /**
+     * Stores the input buffer of generated events in persistent storage and clears it
+     */
+    private func persist_buffer() {
+        storage_manager?.set_persisted("epc_input_buffer", input_buffer)
+    }
+
+    /**
+     * Retrieves persisted input buffer and deletes it from storage
+     *
+     * Merges retrieved events with any existing events in `input_buffer`.
+     */
+    private func recall_buffer() {
+        storage_manager?.get_persisted("epc_input_buffer", completion: {
+            value in
+            if value != nil {
+                let ib = value as! [(stream: String, schema: String, domain: String, data: [String: Any])]
+                for e in ib {
+                    input_buffer.append(e)
+                }
+                storage_manager!.del_persisted("epc_input_buffer")
+            }
+        })
+    }
+
+    /**
+     * This method is called by the application delegate in `applicationWillResignActive()` and
+     * disables event logging.
+     */
+    public func app_in_background() {
+        ENABLED = false
```

Comment:
Do you think we should set `LAST_TS` here as well? It would help match the `session-timed-out()` business logic more I think.

Reply:
Makes sense!